File: game/environment/Tile.cpp
//
// Created by Per-Arne on 24.02.2017.
//

#include <iostream>
#include <SFML/Graphics.hpp>
#include "Tile.h"
#include "../player/Player.h"
#include "../unit/Unit.h"
#include <utility>

Tile::Tile(int x, int y, int width, int height, Tilemap& tilemap):
x(x), y(y), height(height), width(width), tilemap(tilemap)
{

}


const sf::Vector2f Tile::getPixelPosition() {

    return sf::Vector2f(vertices->position.x, vertices->position.y);
}

bool Tile::isAttackable(Unit &unit) {
    if (!occupant)
        return false;

    return occupant->player_.getId() != unit.player_.getId();

}

bool Tile::isWalkable() {
    return walkable;
}

bool Tile::isHarvestable() {
    return harvestable;
}

void Tile::setOccupant(Unit &unit) {
    occupant = &unit;
}

bool Tile::canWalkTo() {
    return occupant == 0 and walkable;
}

bool Tile::isBuildable() {
    return canWalkTo();
}


int Tile::distance(Tile *pTile) {
    return abs(pTile->x - x) + abs(pTile->y - y);
}

void Tile::setDepleted() {
    this->harvestable = false;
    this->walkable = true;

    tilemap.addTileVertices(depleteTile-1, width, height, tilemap.tFirstGid, *this);

}

File: python/game/environment/Tile.py


class Tile:

    def __init__(self, tilemap, x, y, w, h):
        self.id = None
        self.tile_id = None

        self.name = None

        self.tilemap = tilemap
        self.x = x
        self.y = y
        self.width = w
        self.height = h

        self.walkable = False
        self.harvestable = False
        self.swimable = False

        self.resources = 0
        self.oil_yield = 0
        self.lumber_yield = 0
        self.gold_yield = 0

        self.deplete_tile = None